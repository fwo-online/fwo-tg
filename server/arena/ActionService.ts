import arena from '@/arena';
import type { Action } from '@fwo/shared';

export type ActionKey = keyof typeof arena.actions;

export class ActionService {
  static isAction(maybeAction: string): maybeAction is ActionKey {
    return maybeAction in arena.actions;
  }

  static isBaseAction(action: string) {
    if (ActionService.isAction(action)) {
      return arena.actions[action].lvl === 0;
    }

    return false;
  }

  static toObject(actionKey: ActionKey, lvl = 1): Action {
    const action = arena.actions[actionKey];

    const actionObject: Action = {
      name: action.name,
      displayName: action.displayName,
      orderType: action.orderType,
    };

    if ('cost' in action) {
      actionObject.cost = Array.isArray(action.cost) ? action.cost[lvl - 1] : action.cost;
    }

    if ('costType' in action) {
      actionObject.costType = action.costType;
    }

    if ('proc' in action) {
      actionObject.power = action.proc;
    }

    return actionObject;
  }
}
