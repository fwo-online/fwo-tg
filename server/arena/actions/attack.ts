import type { Affect } from '../Constuructors/interfaces/Affect';
import PhysConstructor from '../Constuructors/PhysConstructor';
import CastError from '../errors/CastError';
/**
 * Физическая атака
 */
class Attack extends PhysConstructor implements Affect {
  constructor() {
    super({
      name: 'attack',
      displayName: 'Атака',
      desc: 'Атака',
      lvl: 0,
      orderType: 'all',
    });
  }

  /**
 * Кастомный обработчик атаки
 */
  run() {
    if (!this.params) {
      return;
    }
    const { initiator, target } = this.params;

    target.flags.isHited = {
      initiator: initiator.nick, val: this.status.effect,
    };

    target.stats.down('hp', this.status.effect);
  }

  preAffect: Affect['preAffect'] = ({ params: { initiator, target, game } }): undefined => {
    if (target.flags.isHited) {
      throw new CastError(this.getSuccessResult({ initiator: target, target: initiator, game }));
    }
  };
}
export default new Attack();
