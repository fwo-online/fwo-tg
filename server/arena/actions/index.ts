export { default as attack } from './attack';
export { default as handsHeal } from './handsHeal';
export { default as protect } from './protect';
export { default as regeneration } from './regeneration';
