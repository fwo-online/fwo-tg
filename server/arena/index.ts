import type { Clan } from '@/models/clan';
import type { ItemOutput } from '@fwo/shared';
import type { ItemSet } from '@fwo/shared';
import type { CharacterService } from './CharacterService';
import type GameService from './GameService';
import type MatchMaking from './MatchMakingService';
import type * as actions from './actions';
import type * as magics from './magics';
import type * as passiveSkills from './passiveSkills';
import type * as skills from './skills';
import type * as weaponMastery from './weaponMastery';

export default {
  mm: null as unknown as typeof MatchMaking,
  characters: {} as Record<string, CharacterService>,
  games: {} as Partial<Record<string, GameService>>,
  items: {} as unknown as Record<string, ItemOutput>,
  itemsSets: {} as unknown as Record<string, ItemSet>,
  magics: {} as unknown as typeof magics,
  skills: {} as unknown as typeof skills,
  actions: {} as unknown as typeof actions &
    typeof magics &
    typeof skills &
    typeof passiveSkills &
    typeof weaponMastery,
  clans: new Map<string, Clan>(),
};
