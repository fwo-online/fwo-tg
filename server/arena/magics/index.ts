/**
 * Первичный загрузчик магий
 */
export { default as acidSpittle } from './acidSpittle';
export { default as anathema } from './anathema';
export { default as blessing } from './blessing';
export { default as blight } from './blight';
export { default as bodySpirit } from './bodySpirit';
export { default as chainLightning } from './chainLightning';
export { default as curse } from './curse';
export { default as dispel } from './dispel';
export { default as dustShield } from './dustShield';
export { default as eclipse } from './eclipse';
export { default as entangle } from './entangle';
export { default as exorcism } from './exorcism';
export { default as fireBall } from './fireBall';
export { default as frostTouch } from './frostTouch';
export { default as fireRain } from './fireRain';
export { default as glitch } from './glitch';
export { lightHeal } from './lightHeal';
export { default as lightShield } from './lightShield';
export { default as madness } from './madness';
export { default as magicArmor } from './magicArmor';
export { default as magicArrow } from './magicArrow';
export { default as magicDefense } from './magicDefense';
export { default as mediumAura } from './mediumAura';
export { default as mediumHeal } from './mediumHeal';
export { default as magicWall } from './magicWall';
export { default as paralysis } from './paralysis';
export { default as poisonBreath } from './poisonBreath';
export { default as physicalSadness } from './physicalSadness';
export { default as rockfall } from './rockfall';
export { default as silence } from './silence';
export { default as smallAura } from './smallAura';
export { default as stoneSkin } from './stoneSkin';
export { default as strongHeal } from './strongHeal';
export { default as secondLife } from './secondLife';
export { default as sleep } from './sleep';
export { default as strongAura } from './strongAura';
export { default as vampirism } from './vampirism';
