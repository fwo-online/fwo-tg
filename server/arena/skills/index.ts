export { default as beastCall } from './beastCall';
export { default as berserk } from './berserk';
export { default as disarm } from './disarm';
export { default as dodge } from './dodge';
export { default as parry } from './parry';
export { default as shieldBlock } from './shieldBlock';
export { default as terrifyingHowl } from './terrifyingHowl';
