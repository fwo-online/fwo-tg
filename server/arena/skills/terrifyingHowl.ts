import { shuffle } from 'es-toolkit';
import type { SuccessArgs } from '@/arena/Constuructors/types';
import { joinWithAnd } from '@/arena/LogService/utils/join-with-and';
import { bold, italic } from '@/utils/formatString';
import { Skill } from '../Constuructors/SkillConstructor';

/**
 * üê∫üîä –£–∂–∞—Å–∞—é—â–∏–π –≤–æ–π
 */
class TerrifyingHowl extends Skill {
  constructor() {
    super({
      name: 'terrifyingHowl',
      displayName: 'üê∫üîä –£–∂–∞—Å–∞—é—â–∏–π –≤–æ–π',
      desc: '–ó–∞—Å—Ç–∞–≤–ª—è–µ—Ç –≤—Ä–∞–≥–æ–≤ –æ—Ü–µ–ø–µ–Ω–µ—Ç—å',
      cost: [50, 75, 100],
      proc: 50,
      baseExp: 8,
      costType: 'en',
      orderType: 'enemy',
      aoeType: 'target',
      chance: [80, 90, 100],
      effect: [0, 1, 3],
      profList: {},
      bonusCost: [0, 0, 0],
    });
  }

  run() {
    const { initiator, target, game } = this.params;
    const initiatorMagicLvl = initiator.skills[this.name];
    const effect = this.effect[initiatorMagicLvl - 1] || 1;

    const enemies = shuffle(game.players.getAliveEnemies(initiator))
      .filter(({ id }) => id !== target.id)
      .splice(0, effect)
      .concat(target);

    enemies.forEach((target) => {
      target.flags.isParalysed.push({ initiator, val: effect });
      this.status.expArr.push({ initiator, target });
    });

    this.calculateExp();
  }

  customMessage(args: SuccessArgs) {
    return `${bold(args.initiator.nick)} –∏—Å–ø—É—Å—Ç–∏–ª ${italic(this.displayName)}! ${joinWithAnd(args.expArr.map(({ target }) => bold(target.nick)))} –Ω–µ –≤ —Å–∏–ª–∞—Ö –¥–≤–∏–≥–∞—Ç—å—Å—è –æ—Ç —Å—Ç—Ä–∞—Ö–∞!`;
  }
}

export default new TerrifyingHowl();
