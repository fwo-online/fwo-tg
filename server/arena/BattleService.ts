import { Markup } from 'telegraf';
import BattleKeyboard from '@/helpers/BattleKeyboard';
import * as channelHelper from '../helpers/channelHelper';
import OrderError from './errors/OrderError';
import type Game from './GameService';
import arena from './index';

type BattleReply = {
  message: string;
  keyboard: ReturnType<typeof Markup.button.callback>[][];
}

/**
 * –ó–∞–∫–∞–∑—ã–≤–∞–µ–º –≤ —Ç—Ä–∏ —ç—Ç–∞–ø–∞:
 * 1. –í—ã–±–æ—Ä —É–º–µ–Ω–∏—è (action_{attack})
 * 2. –í—ã–±–æ—Ä —Ü–µ–ª–∏ ({attack}_{target})
 * 3. –í—ã–±–æ—Ä —Å–∏–ª—ã ({attack}_{target}_{proc})
 */

const getActions = () => ({ ...arena.actions, ...arena.skills, ...arena.magics });

const getText = {
  order: (proc: number) => `–£ —Ç–µ–±—è –æ—Å—Ç–∞–ª–æ—Å—å *${proc}%*
–ó–∞–∫–∞–∑—ã:\n`,
  proc: (displayName: string, target: string) => `–í—ã–±–µ—Ä–∏ —Å–∏–ª—É _${displayName}_ –Ω–∞ ${target ? `–∏–≥—Ä–æ–∫–∞ ${target}` : '—Å–µ–±—è'}`,
  target: (displayName: string) => `–í—ã–±–µ—Ä–∏ —Ü–µ–ª—å –¥–ª—è _${displayName}_`,
};

function getCurrentOrders(charId: string, game: Game) {
  return game.orders.getPlayerOrders(charId)
    .map((o) => {
      if (o.target === charId) {
        return `_${getActions()[o.action].displayName}_ (*${o.proc}%*) –Ω–∞ *—Å–µ–±—è*`;
      }
      const target = game.players.getById(o.target);
      if (!target) {
        throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
      }
      return `_${getActions()[o.action].displayName}_ (*${o.proc}%*) –Ω–∞ –∏–≥—Ä–æ–∫–∞ *${target.nick}*`;
    })
    .join('\n');
}

function getTargetKeyboard(charId: string, game: Game, action: string) {
  const player = game.players.getById(charId);
  if (!player) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const orders = game.orders.getPlayerOrders(charId);
  const { orderType } = getActions()[action];
  const proc = arena.skills[action] ? `_${arena.skills[action].proc}` : '';

  return game.players.alivePlayers
    .filter((target) => {
      if (orderType === 'enemy') {
        return !player.isAlly(target);
      }
      if (orderType === 'team') {
        return player.isAlly(target);
      }
      if (orderType === 'teamExceptSelf') {
        return player.isAlly(target, false);
      }

      return !orders.some((order) => target.id === order.target && action === order.action);
    })
    .map(({ nick, id }) => Markup.button.callback(nick, `${action}_${id}${proc}`));
}

/**
 * –í–æ–∑–≤—Ä–∞—â–∞–µ—Ç –∫–Ω–æ–ø–∫–∏ —Å –ø—Ä–æ—Ü–µ–Ω—Ç–æ–º –∑–∞–∫–∞–∑–∞
 */
function getPercentKeyboard(action: string, target: string, proc: number) {
  return Array
    .from(new Set([5, 10, 25, 50, 75, proc]))
    .filter((key) => key <= proc)
    .map((key) => Markup.button.callback(key.toString(), `${action}_${target}_${key}`));
}

/**
 * –°–æ–æ–±—â–µ–Ω–∏–µ –¥–ª—è –ø–µ—Ä–≤–æ–≥–æ —ç—Ç–∞–ø–∞
 * @param charId
 * @param game
 */
function orderMessage(charId: string, game: Game): BattleReply {
  const player = game.players.getById(charId);
  if (!player) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const message = getText.order(player.proc)
    .concat(getCurrentOrders(charId, game));
  const keyboard = channelHelper.getOrderButtons(player);
  return { message, keyboard };
}

/**
 * –°–æ–æ–±—â–µ–Ω–∏–µ –¥–ª—è –≤—Ç–æ—Ä–æ–≥–æ —ç—Ç–∞–ø–∞
 */
function targetMessage(charId: string, game: Game, action: string): BattleReply {
  const { displayName } = getActions()[action];
  const message = getText.target(displayName);
  const keyboard = [
    getTargetKeyboard(charId, game, action),
    [Markup.button.callback('üîô –ù–∞–∑–∞–¥', 'back')],
  ];
  return { message, keyboard };
}

type PercentMessageParams = {
  initiator: string;
  game: Game;
  action: string;
  target: string;
  isSelfAction?: boolean;
}
/**
 * –°–æ–æ–±—â–µ–Ω–∏–µ –¥–ª—è —Ç—Ä–µ—Ç—å–µ–≥–æ —ç—Ç–∞–ø–∞
 */
function percentMessage({
  initiator: initiatorId, game, action, target: targetId, isSelfAction = false,
}: PercentMessageParams) {
  const initiator = game.players.getById(initiatorId);
  if (!initiator) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const target = game.players.getById(targetId);
  if (!target) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const { displayName } = getActions()[action];
  const backButtonData = isSelfAction ? 'back' : `action_${action}`;
  const message = getText.proc(displayName, isSelfAction ? '' : target.nick);
  const keyboard = [
    getPercentKeyboard(action, targetId, initiator.proc),
    [Markup.button.callback('üîô –ù–∞–∑–∞–¥', backButtonData)],
  ];
  return { message, keyboard };
}

export function getDefaultMessage(charId: string, game: Game): BattleReply {
  return orderMessage(charId, game);
}

export function getAllMagicsMessage(charId: string, game: Game): BattleReply {
  const player = game.players.getById(charId);
  if (!player) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const message = '–°–ø–∏—Å–æ–∫ –≤—Å–µ—Ö –º–∞–≥–∏–π';

  const keyboard = [
    ...new BattleKeyboard(player).setMagics(true).render(true),
    [Markup.button.callback('–ù–∞–∑–∞–¥', 'back')],
  ];

  return { message, keyboard };
}
/**
   * –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤—ã–±—Ä–∞–Ω–Ω–æ–≥–æ –¥–µ–π—Å—Ç–≤–∏—è (–ø–µ—Ä–≤—ã–π —ç—Ç–∞–ø)
   */
export function handleAction(initiator: string, game: Game, action: string): BattleReply {
  if (action === 'repeat') return repeatOrder(initiator, game);
  if (action === 'reset') return resetOrder(initiator, game);

  const { orderType } = getActions()[action];
  const proc = arena.skills[action] ? arena.skills[action].proc : undefined;

  // –ï—Å–ª–∏ —É–º–µ–Ω–∏–µ –Ω–∞–ø—Ä–∞–≤–ª–µ–Ω–æ –Ω–∞ —Å–µ–±—è –∏ –∏–º–µ–µ—Ç –ø—Ä–æ—Ü–µ–Ω—Ç (—É–º–µ–Ω–∏—è), —Ç–æ –¥–µ–ª–∞–µ–º –∑–∞–∫–∞–∑ —Å—Ä–∞–∑—É
  if (orderType === 'self') {
    if (typeof proc !== 'undefined') {
      game.orders.orderAction({
        initiator, target: initiator, action, proc,
      });
      return orderMessage(initiator, game);
    }
    return percentMessage({
      initiator, game, action, target: initiator, isSelfAction: true,
    });
  }
  return targetMessage(initiator, game, action);
}
/**
 * –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤—ã–±—Ä–∞–Ω–Ω–æ–π —Ü–µ–ª–∏ (–≤—Ç–æ—Ä–æ–π —ç—Ç–∞–ø)
 */
// eslint-disable-next-line max-len
export function handleTarget(initiator: string, game: Game, action: string, target: string): BattleReply {
  const proc = arena.skills[action] ? arena.skills[action].proc : undefined;

  if (typeof proc !== 'undefined') {
    game.orders.orderAction({
      initiator, target, action, proc,
    });
    return orderMessage(initiator, game);
  }
  return percentMessage({
    initiator, game, action, target,
  });
}
/**
 * –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤—ã–±—Ä–∞–Ω–Ω—ã—Ö –ø—Ä–æ—Ü–µ–Ω—Ç–æ–≤ (—Ç—Ä–µ—Ç–∏–π —ç—Ç–∞–ø)
 */
// eslint-disable-next-line max-len
export function handlePercent(charId: string, game: Game, action: string, target: string, proc: number): BattleReply {
  game.orders.orderAction({
    initiator: charId, target, action, proc,
  });
  return orderMessage(charId, game);
}

export function repeatOrder(charId: string, game: Game): BattleReply {
  game.orders.repeatLastOrder(charId);
  return orderMessage(charId, game);
}

export function resetOrder(charId: string, game: Game): BattleReply {
  game.orders.resetOrdersForPlayer(charId);
  return orderMessage(charId, game);
}
