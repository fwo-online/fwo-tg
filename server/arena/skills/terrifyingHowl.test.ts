import { afterEach, beforeEach, describe, expect, it } from 'bun:test';
import { CharacterClass } from '@fwo/shared';
import type GameService from '@/arena/GameService';
import { eclipse, paralysis } from '@/arena/magics';
import TestUtils from '@/utils/testUtils';
import attack from '../actions/attack';
import terrifyingHowl from './terrifyingHowl';

// npm t server/arena/skills/terrifyingHowl.test.ts

describe('terrifyingHowl', () => {
  let game: GameService;

  beforeEach(async () => {
    attack.registerPreAffects([paralysis]);
    eclipse.registerPreAffects([paralysis]);

    game = await TestUtils.createGame([
      {
        prof: CharacterClass.Warrior,
        skills: { terrifyingHowl: 2 },
      },
      {
        prof: CharacterClass.Mage,
        magics: { eclipse: 1 },
      },
      {},
    ]);

    TestUtils.mockRandom();
  });

  afterEach(() => {
    TestUtils.restoreRandom();
  });

  it('should apply paralysis', async () => {
    game.players.players[0].proc = 1;
    game.players.players[1].proc = 1;
    game.players.players[0].stats.set('en', 999);
    game.players.players[1].stats.set('mp', 999);
    terrifyingHowl.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players[1].flags.isParalysed).toHaveLength(1);
    expect(game.players.players[2].flags.isParalysed).toHaveLength(1);

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
