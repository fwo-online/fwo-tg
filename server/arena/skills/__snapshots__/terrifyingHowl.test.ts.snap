// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`terrifyingHowl should apply paralysis 1`] = `
"*asperiores* испустил _🐺🔊 Ужасающий вой_! *ratione* и *alias* не в силах двигаться от страха!
\\[ 📖8 ]"
`;
